ExtractUserDataActivity/handler.ts

```diff
@@ -0,0 +1,526 @@
+      () => profileModel.findOneProfileByFiscalCode(fiscalCode),
+      "findOneProfileByFiscalCode"
+    ).foldTaskEither<
+      ActivityResultUserNotFound | ActivityResultQueryFailure,
+      Profile
+    >(
+      failure => fromEither(left(failure)),
+      maybeProfile =>
+        fromEither<ActivityResultUserNotFound, Profile>(
+          fromOption(
+            ActivityResultUserNotFound.encode({
+              kind: "USER_NOT_FOUND_FAILURE"
+            })
+          )(maybeProfile)
+        )
+    );
+
+  /**
+   * Given a message as it's retrieved from the database, it queries the blob storage for the content and returns a new version of the message including it
+   * @param messageFromDb a message as it's retrieved from db
+   * @return either the message with it's content or a query error
+   */
+  const getMessageWithContent = (
+    messageFromDb: RetrievedMessageWithoutContent
+  ): TaskEither<ActivityResultQueryFailure, MessageWithContent> => {
+    return fromQueryEither(
+      () => messageModel.getContentFromBlob(blobService, messageFromDb.id),
+      "messageModel.getContentFromBlob"
+    ).foldTaskEither<ActivityResultQueryFailure, MessageWithContent>(
+      failure => fromEither(left(failure)),
+      maybeContent =>
+        fromEither<ActivityResultQueryFailure, MessageContent>(
+          fromOption(
+            ActivityResultQueryFailure.encode({
+              kind: "QUERY_FAILURE",
+              query: "messageModel.getContentFromBlob",
+              reason: `Cannot find content for message ${messageFromDb.id}`
+            })
+          )(maybeContent)
+        ).map<MessageWithContent>((content: MessageContent) =>
+          appendContentToMessage(messageFromDb, content)
+        )
+    );
+  };
+
+  /**
+   * Utility that performs enrichMessage over a list of messages
+   * @param messages a list of messages without content
+   * @returns either a list of message with content or a query error
+   */
+  const getAllMessagesWithContent = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<MessageWithContent>
+  > => {
+    if (messages.length) {
+      // this spread is needed as typescript wouldn't recognize messages[0] to be defined otherwise
+      const [firstMessage, ...otherMessages] = messages;
+      return sequenceT(taskEither)(
+        getMessageWithContent(firstMessage),
+        ...otherMessages.map(getMessageWithContent)
+      );
+    }
+    return taskEither.of([]);
+  };
+
+  /**
+   * Given a list of messages, it queires for relative notifications
+   * @param messages
+   */
+  const findNotificationsForAllMessages = (
+    messages: readonly RetrievedMessageWithoutContent[]
+  ): TaskEither<
+    ActivityResultQueryFailure,
+    ReadonlyArray<RetrievedNotification>
+  > => {
+    if (messages.length) {
+      // this spread is needed as typescript wouldn't recognize messages[0] to be defined otherwise
```

Comment:
it's not clear, why do we need this pattern here ? consider using a NonEmptyArray instead

NonEmptyArray.decode(messages)... or if (NonEmptyArray.is(message)) ...

Reply:
"fp-ts/lib/NonEmptyArray"

Reply:
didn't know that, thanks

Reply:
are you sure it works like that? it doesn't seem to be a decoder though.

Anyway, I think it doesn't solved the issue: to pass an arbitrary list of taskEither to a `sequenceT` to resolve them in parallel: https://github.com/pagopa/io-functions-admin/pull/42/files/98dd1f4ddb37bed424eb078bd0549cd0a9c064b9#diff-6c3441a7b004efdd3512746300508009R339

Reply:
I mean: you can just use `NonEmptyArray.fromArray(messages).map(...)` instead of check messages length and separate head and tail (it's not clear why do you want to do that)